### DISCUSSION TITLE
fireurl hardening

### DISCUSSION THREAD

**Discussion by topimiettinen** [rusty-snake/fireurl#4] [Ideas]
Fireurld listens to a Unix socket in $XDG_RUNTIME_DIR, reads a URL string from the socket and executes `firefox` with URL as argument. There's a check for validity of UTF-8 in the URL.

Fireurld could be hardened:
- reject too short URLs
- IDK how Rust's `Command::new` works, but if it uses a shell, special characters used by shells like `!"#$%&()[]?|;` should be rejected, also NUL/CR/LF
- socket mode 0600
- timeout if reading takes too long
- enforce a maximum number of simultaneous connection attempts
- check client credentials with SO_PEERCRED & SO_PEERSEC (useful to also filter clients by SELinux label)
- authenticate clients: pass a secret with `keyctl` or less securely with env variable: launch also client with fireurld?
- per-client URL filtering at server side (for example, reject `file:` or `about:` URLs or only allow `https:`, regexp/fnmatch patterns: `allow ^https://.*\.example.com$`)

The command should be configurable to use chrome instead of FF and for flags like `--incognito` etc.

WDYT?

**Comment by rusty-snake** [comment]
> There's a check for validity of UTF-8 in the URL.

This check originates in the fact parsing and manipulating strings in Rust is much easier with `str` (requires utf-8) than with `OsStr` (just bytes).

> reject too short URLs

When is it too short?

> IDK how Rust's Command::new works, but if it uses a shell, special characters used by shells like !"#$%&()[]?|; should be rejected, also NUL/CR/LF

Think of it like `execvpe` i.e. no shell.

NUL-bytes will cause a `NulError` in Rust.

> The command should be configurable to use chrome instead of FF and for flags like --incognito etc.

See #2 for my idea.

> socket mode 0600

`chmod` after the `bind`?

> timeout if reading takes too long

~What is a good timeout in your eyes? one second?~ Actually the `recv` is the blocking syscall we do in the main loop to wait for input. Setting a timeout here would mean that we will produce errors all the time unless you open links in that frequency.

> launch also client with fireurld?

Since clients (fireurl) are expected to be started in a sandbox, how would this look like?

I have also already thought about linking the requests to sandboxes (and unsandboxed). However, I have not yet found a reliable (=secure) solution for this.

**Reply by topimiettinen** [reply]
> > reject too short URLs
> 
> When is it too short?

Zero length definitely, probably 1 to 3 too.

> > IDK how Rust's Command::new works, but if it uses a shell, special characters used by shells like !"#$%&()[]?|; should be rejected, also NUL/CR/LF
> 
> Think of it like `execvpe` i.e. no shell.
> 
> NUL-bytes will cause a `NulError` in Rust.

Perhaps allow-list approach should be used and only allow valid URL characters from RFC 1738.

> > The command should be configurable to use chrome instead of FF and for flags like --incognito etc.
> 
> See #2 for my idea.

Nice, then for example mailto: URLs can use a different program.

**Comment by rusty-snake** [comment]
Also partial reads and writes need to be checked.
